File: scripts/node/parsers/reportParser.js
'use strict';

const fs   = require('fs');
const path = require('path');

function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

/**
 * Reads Report metadata from SFDX source format (reports are nested in folders).
 * @param {string} typeDir - path to the reports/ folder
 * @returns {{ reports: Array }}
 */
async function parseReports(typeDir) {
  const reports = [];
  if (!fs.existsSync(typeDir)) return { reports };

  const walk = (dir) => {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(full);
      } else if (entry.name.endsWith('.report-meta.xml')) {
        const xml = fs.readFileSync(full, 'utf8');
        reports.push({
          name:        entry.name.replace('.report-meta.xml', ''),
          reportType:  extractTag(xml, 'reportType'),
          format:      extractTag(xml, 'format'),
          description: extractTag(xml, 'description'),
        });
      }
    }
  };

  walk(typeDir);
  return { reports };
}

module.exports = { parseReports };

File: scripts/python/parsers/report_parser.py
"""Parser for Report metadata (.report-meta.xml), nested in folder hierarchy."""
import re
from pathlib import Path


def _extract_tag(xml: str, tag: str) -> str | None:
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def parse_reports(type_dir: str) -> dict:
    """
    @param type_dir: path to the reports/ folder (reports are nested in sub-folders)
    @returns: { reports: list }
    """
    reports = []
    p = Path(type_dir)
    if not p.exists():
        return {'reports': reports}

    for f in sorted(p.rglob('*.report-meta.xml')):
        xml = f.read_text(encoding='utf-8')
        reports.append({
            'name':        f.name.replace('.report-meta.xml', ''),
            'reportType':  _extract_tag(xml, 'reportType'),
            'format':      _extract_tag(xml, 'format'),
            'description': _extract_tag(xml, 'description'),
        })

    return {'reports': reports}
